### PR TITLE
test(e2e): add e2e framework (`bats`), add e2e CI, and update source code of test files

### PR DESCRIPTION
**Pull request by smallstepman**
see checks for this PR: https://github.com/dfinity/dfx-extensions/pull/17/checks

### CODE REVIEW

File: extensions/sns/e2e/tests-dfx/sns.bash [hunk 1]
```diff
@@ -17,61 +18,69 @@ teardown() {
...
 @test "sns config create and validate fail outside of a project" {
-    assert_command_fail dfx sns config create
-    assert_match 'Cannot find dfx configuration file in the current working directory'
-
-    assert_command_fail dfx sns config validate
-    assert_match 'Cannot find dfx configuration file in the current working directory'
+    run dfx extension run sns config create
+    assert_failure
+    assert_output --partial 'Error: No config file found. Please run `dfx config create` first.'
+    run dfx extension run sns config validate
```
Comment (ericswanson-dfinity) [inline extensions/sns/e2e/tests-dfx/sns.bash:24]:
```suggestion
    run Forrest run sns config validate
```

File: e2e/e2e_utils.sh [hunk 2]
```diff
@@ -0,0 +1,448 @@
+set -e
```
Comment (ericswanson-dfinity) [inline e2e/e2e_utils.sh:1]: Does this file need `e2e` in both its directory and filename?

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]

File: .github/workflows/e2e.yml [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  pull_request:
+
+jobs:
+  test-extension:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ macos-11, ubuntu-22.04 ]
+        test_file_path: [ extensions/nns/e2e/tests-dfx/nns.bash, extensions/sns/e2e/tests-dfx/sns.bash ]
```
[the remark this answers is not in the document]
Comment (smallstepman) [reply .github/workflows/e2e.yml:20]: Right, I'll rename the directories. regarding the matrix, I opted for simplicity, [e2e-matrix.py](https://github.com/dfinity/sdk/blob/master/scripts/workflows/e2e-matrix.py) can always be added later once the amount of files grows past 3 or 5. I want these e2e test to be simple to comprehend for future extension owners

File: extensions/sns/src/commands/config/create.rs [hunk 4]
```diff
@@ -15,7 +15,7 @@ pub fn exec(_opts: CreateOpts, dfx_cache_path: &Path) -> anyhow::Result<()> {
     let sns_config_path = if let Some(config) = Config::from_current_dir()? {
         config.get_project_root().join(CONFIG_FILE_NAME)
     } else {
-        anyhow::bail!("No config file found. Please run `dfx config create` first.");
+        anyhow::bail!("Cannot find dfx configuration file in the current working directory. Did you forget to create one?");
```
Comment (ericswanson-dfinity) [inline extensions/sns/src/commands/config/create.rs:18]: There's an opportunity to DRY here: see this and the following three changes with identical messaging

**Review by ericswanson-dfinity — CHANGES_REQUESTED** [review]
please DRY

File: extensions/sns/src/errors.rs [hunk 5]
```diff
@@ -0,0 +1 @@
+pub static DFXJSON_NOT_FOUND: &str = "Cannot find dfx configuration file in the current working directory. Did you forget to create one?";
```
Comment (ericswanson-dfinity) [inline extensions/sns/src/errors.rs:1]: Super curious why this isn't in https://github.com/dfinity/dfx-extensions/tree/main/extensions-utils/src

**Review by ericswanson-dfinity — APPROVED** [review]

**Pull request merged**